kubernetes-itests/src/test/java/io/fabric8/openshift/UserImpersonationIT.java

```diff
@@ -0,0 +1,145 @@
+
+  private ServiceAccount serviceAccount1;
+  private KubernetesClusterRole impersonatorRole;
+  private KubernetesClusterRoleBinding impersonatorRoleBinding;
+
+  private String currentNamespace;
+
+  @Before
+  public void init() {
+    currentNamespace = session.getNamespace();
+    // Create impersonator cluster role
+    impersonatorRole = new KubernetesClusterRoleBuilder()
+      .withNewMetadata()
+      .withName("impersonator")
+      .endMetadata()
+      .addToRules(new KubernetesPolicyRuleBuilder()
+        .addToApiGroups("")
+        .addToResources("users", "groups", "serviceaccounts")
+        .addToVerbs("impersonate")
+        .build()
+      )
+      .build();
+    client.rbac().kubernetesClusterRoles().inNamespace(currentNamespace).createOrReplace(impersonatorRole);
+
+    // Create Service Account
+    serviceAccount1 = new ServiceAccountBuilder()
+      .withNewMetadata().withName(SERVICE_ACCOUNT).endMetadata()
+      .build();
+    client.serviceAccounts().inNamespace(currentNamespace).create(serviceAccount1);
+
+    // Bind Impersonator Role to current user
+    impersonatorRoleBinding = new KubernetesClusterRoleBindingBuilder()
+      .withNewMetadata()
+      .withName("impersonate-role")
+      .endMetadata()
+      .addToSubjects(new KubernetesSubjectBuilder()
+        .withApiGroup("rbac.authorization.k8s.io")
+        .withKind("User")
+        .withName(client.currentUser().getMetadata().getName())
+        .withNamespace(currentNamespace)
+        .build()
+      )
+      .withRoleRef(new KubernetesRoleRefBuilder()
+        .withApiGroup("rbac.authorization.k8s.io")
+        .withKind("ClusterRole")
+        .withName("impersonator")
+        .build()
+      )
+      .build();
+
+    client.rbac().kubernetesClusterRoleBindings().inNamespace(currentNamespace).createOrReplace(impersonatorRoleBinding);
+  }
+
+
+  @Test
+  public void testUserImpersonation() {
+    RequestConfig requestConfig = client.getConfiguration().getRequestConfig();
+    requestConfig.setImpersonateUsername(SERVICE_ACCOUNT);
+    requestConfig.setImpersonateGroups("system:authenticated", "system:authenticated:oauth");
+
+    User user = client.currentUser();
+    assertThat(user.getMetadata().getName()).isEqualTo(SERVICE_ACCOUNT);
+  }
+
+
+  @Test
+  public void testCreateProjectImpersonatingServiceAccount() {
+    RequestConfig requestConfig = client.getConfiguration().getRequestConfig();
+    requestConfig.setImpersonateUsername(SERVICE_ACCOUNT);
+    requestConfig.setImpersonateGroups("system:authenticated", "system:authenticated:oauth");
+    // Create a project
+    ProjectRequest projectRequest = client.projectrequests().createNew()
+      .withNewMetadata()
+      .withName(NEW_PROJECT)
+      .endMetadata()
+      .done();
+
+    // Grab the requester annotation
+    String requester = projectRequest.getMetadata().getAnnotations().get("openshift.io/requester");
+    assertThat(requester).isEqualTo(SERVICE_ACCOUNT);
```

Comment:
I don't grok how in fact we are testing that we can impersonate groups set in the request? Is that a matter of better test description or we should make different assertions here?

Reply:
There are only two groups in this case: https://docs.openshift.com/container-platform/3.10/architecture/additional_concepts/authentication.html#authentication-impersonation

Reply:
Ok, so what you are testing here is if you can create a project impersonating other user with those groups. Then maybe the test name can be more descriptive. WDYT?

Reply:
Any suggestions on the test name? `testCreateProjectImpersonatingServiceAccount` seemed pretty descriptive to me :smiley:

Reply:
groups sth sth

Reply:
"I should be able to create a project impersonating user while using service account and groups" maybe

Reply:
Done